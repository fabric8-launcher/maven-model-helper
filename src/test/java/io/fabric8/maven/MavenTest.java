package io.fabric8.maven;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.StringReader;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.Properties;

import org.apache.maven.model.Dependency;
import org.apache.maven.model.Model;
import org.apache.maven.model.Scm;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.xmlunit.assertj.XmlAssert;

/**
 * @author <a href="mailto:ggastald@redhat.com">George Gastaldi</a>
 */
class MavenTest {

    @Test
    void should_read_model() {
        Path basePom = Paths.get("pom.xml");
        Model model = Maven.readModel(basePom);
        Assertions.assertThat(model).isNotNull();
        assertThat(model.getPomFile().getAbsolutePath()).isEqualTo(basePom.toAbsolutePath().toString());
        assertThat(model.getParent().getGroupId()).isEqualTo("org.jboss");
        assertThat(model.getArtifactId()).isEqualTo("maven-model-helper");
    }

    @Test
    void should_read_model_using_reader() {
        Model model = Maven.readModel(
                new StringReader("<project><groupId>org.jboss</groupId><artifactId>maven-model-helper</artifactId></project>"));
        Assertions.assertThat(model).isNotNull();
        assertThat(model.getGroupId()).isEqualTo("org.jboss");
        assertThat(model.getArtifactId()).isEqualTo("maven-model-helper");
    }

    @Test
    void should_read_model_string() {
        Path basePom = Paths.get("pom.xml");
        Model model = Maven.readModel(basePom.toAbsolutePath().toString());
        Assertions.assertThat(model).isNotNull();
        assertThat(model.getPomFile().getAbsolutePath()).isEqualTo(basePom.toAbsolutePath().toString());
        assertThat(model.getParent().getGroupId()).isEqualTo("org.jboss");
        assertThat(model.getArtifactId()).isEqualTo("maven-model-helper");
    }

    @Test
    void should_write_model(@TempDir Path tempDir) throws IOException {
        File pom = tempDir.resolve("temp-pom.xml").toFile();
        Model model = new Model();
        model.setPomFile(pom);
        model.setGroupId("org.example");
        model.setArtifactId("example");
        model.setVersion("1.0");
        Maven.writeModel(model);
        assertThat(Files.readAllLines(pom.toPath()).stream().map(String::trim))
                .contains("<groupId>org.example</groupId>", "<artifactId>example</artifactId>", "<version>1.0</version>");
    }

    @Test
    void should_write_model_with_sorted_properties(@TempDir Path tempDir) throws IOException {
        Path basePom = Paths.get("pom.xml");
        Model model = Maven.readModel(basePom.toAbsolutePath().toString());

        Properties properties = model.getProperties();
        assertThat(properties).isInstanceOf(SortedProperties.class);

        properties.put("c", "three");
        properties.put("a", "one");
        properties.put("b", "two");

        // Write pom
        Path pom = tempDir.resolve("temp-pom.xml");
        Maven.writeModel(model, pom);
        assertThat(Files.readAllLines(pom).stream().map(String::trim))
                .containsSequence("<a>one</a>", "<b>two</b>", "<c>three</c>");
    }

    @Test
    void should_write_model_with_sorted_properties_using_reader(@TempDir Path tempDir) throws IOException {
        Path basePom = Paths.get("pom.xml");
        Model model = Maven.readModel(new FileReader(basePom.toFile()));

        Properties properties = model.getProperties();
        assertThat(properties).isInstanceOf(SortedProperties.class);

        properties.put("c", "three");
        properties.put("a", "one");
        properties.put("b", "two");

        // Write pom
        Path pom = tempDir.resolve("temp-pom.xml");
        Maven.writeModel(model, pom);
        assertThat(Files.readAllLines(pom).stream().map(String::trim))
                .containsSequence("<a>one</a>", "<b>two</b>", "<c>three</c>");
    }

    @Test
    void should_preserve_parent_relative_path(@TempDir Path tempDir) throws Exception {
        URL resource = getClass().getResource("parent/parent-pom.xml");
        Path parentPom = Paths.get(resource.toURI());
        Path newPath = tempDir.resolve("new-pom.xml");
        Model model = Maven.readModel(parentPom);
        assertThat(model.getParent().getRelativePath()).isNotNull();

        Maven.writeModel(model, newPath);
        XmlAssert.assertThat(newPath)
                .withNamespaceContext(Collections.singletonMap("maven", "http://maven.apache.org/POM/4.0.0"))
                .valueByXPath("//maven:project/maven:parent/maven:relativePath")
                .isEqualTo("../../pom.xml");
    }

    @Test
    void should_write_scm_tag(@TempDir Path tempDir) throws Exception {
        Path pomXml = tempDir.resolve("pom.xml");
        Model newModel = Maven.newModel();
        Scm scm = new Scm();
        scm.setUrl("https://github.com/fabric8-launcher/maven-model-helper.git");
        scm.setDeveloperConnection("scm:git:git@github.com:fabric8-launcher/maven-model-helper.git");
        scm.setTag("HEAD");
        newModel.setScm(scm);
        Maven.writeModel(newModel, pomXml);
        XmlAssert.assertThat(pomXml)
                .withNamespaceContext(Collections.singletonMap("maven", "http://maven.apache.org/POM/4.0.0"))
                .valueByXPath("//maven:project/maven:scm/maven:url")
                .isEqualTo(scm.getUrl());
        XmlAssert.assertThat(pomXml)
                .withNamespaceContext(Collections.singletonMap("maven", "http://maven.apache.org/POM/4.0.0"))
                .valueByXPath("//maven:project/maven:scm/maven:developerConnection")
                .isEqualTo(scm.getDeveloperConnection());
        // HEAD is not written
        XmlAssert.assertThat(pomXml)
                .withNamespaceContext(Collections.singletonMap("maven", "http://maven.apache.org/POM/4.0.0"))
                .valueByXPath("//maven:project/maven:scm/maven:tag")
                .isEmpty();
        ;

    }

    @Test
    void should_respect_insertion_order(@TempDir Path tempDir) throws Exception {
        URL resource = getClass().getResource("parent/parent-pom.xml");
        Path basePom = Paths.get(resource.toURI());
        Model model = Maven.readModel(basePom);

        Path updatedPom = tempDir.resolve("updated-pom.xml");
        Files.copy(basePom, updatedPom);

        Dependency dependency = model.getDependencies().stream().filter(d -> d.getArtifactId().equals("quarkus-junit5-internal")).findFirst().orElseThrow();
        dependency.setVersion("1.0.0");
        dependency.setOptional("true");
        Maven.writeModel(model, updatedPom);

        assertThat(Files.readString(updatedPom))
                .contains("<dependency>\n" +
                        "      <groupId>io.quarkus</groupId>\n" +
                        "      <artifactId>quarkus-junit5-internal</artifactId>\n" +
                        "      <version>1.0.0</version>\n" +
                        "      <scope>test</scope>\n" +
                        "      <optional>true</optional>\n" +
                        "    </dependency>");
    }
}
