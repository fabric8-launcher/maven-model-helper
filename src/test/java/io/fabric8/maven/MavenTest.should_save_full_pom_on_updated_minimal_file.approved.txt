<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <parent>
    <artifactId>jboss-parent</artifactId>
    <groupId>org.jboss</groupId>
    <version>43</version>
  </parent>
  <modelVersion>4.0.0</modelVersion>
  <groupId>org.acme</groupId>
  <artifactId>full-pom</artifactId>
  <packaging>pom</packaging>
  <name>Full Test POM.xml</name>
  <version>0.0.1</version>
  <description>Full pom.xml for testing purpose</description>
  <url>https://acme.org/full-pom</url>
  <prerequisites>
    <maven>string</maven>
  </prerequisites>
  <issueManagement>
    <system>string</system>
    <url>string</url>
  </issueManagement>
  <ciManagement>
    <system>string</system>
    <url>string</url>
    <notifiers>
      <notifier>
        <type>string</type>
        <sendOnSuccess>false</sendOnSuccess>
        <address>string</address>
        <configuration>
          <token>1324567890</token>
        </configuration>
      </notifier>
    </notifiers>
  </ciManagement>
  <inceptionYear>2024</inceptionYear>
  <mailingLists>
    <mailingList>
      <name>string</name>
      <subscribe>string</subscribe>
      <unsubscribe>string</unsubscribe>
      <post>string</post>
      <archive>string</archive>
      <otherArchives>
        <otherArchive>string</otherArchive>
      </otherArchives>
    </mailingList>
  </mailingLists>
  <developers>
    <developer>
      <id>string</id>
      <name>string</name>
      <email>string</email>
      <url>string</url>
      <organization>string</organization>
      <organizationUrl>string</organizationUrl>
      <roles>
        <role>string</role>
      </roles>
      <timezone>string</timezone>
      <properties>
        <country>FRANCE</country>
      </properties>
    </developer>
  </developers>
  <contributors>
    <contributor>
      <name>string</name>
      <email>string</email>
      <url>string</url>
      <organization>string</organization>
      <organizationUrl>string</organizationUrl>
      <roles>
        <role>string</role>
      </roles>
      <timezone>string</timezone>
      <properties>
        <favorite.color>BLUE</favorite.color>
      </properties>
    </contributor>
  </contributors>
  <licenses>
    <license>
      <name>string</name>
      <url>string</url>
      <distribution>string</distribution>
      <comments>string</comments>
    </license>
  </licenses>
  <scm>
    <connection>scm:git:git://github.com/fabric8-launcher/maven-model-helper.git</connection>
    <developerConnection>scm:git:git@github.com:fabric8-launcher/maven-model-helper.git</developerConnection>
    <url>https://github.com/fabric8-launcher/maven-model-helper</url>
  </scm>
  <organization>
    <name>ACME</name>
    <url>https://acme.org</url>
  </organization>
  <build>
    <sourceDirectory>string</sourceDirectory>
    <scriptSourceDirectory>string</scriptSourceDirectory>
    <testSourceDirectory>string</testSourceDirectory>
    <outputDirectory>string</outputDirectory>
    <testOutputDirectory>string</testOutputDirectory>
    <extensions>
      <extension>
        <groupId>string</groupId>
        <artifactId>string</artifactId>
        <version>string</version>
      </extension>
    </extensions>
    <defaultGoal>string</defaultGoal>
    <resources>
      <resource>
        <targetPath>string</targetPath>
        <filtering>true</filtering>
        <directory>string</directory>
        <includes>
          <include>string</include>
        </includes>
        <excludes>
          <exclude>string</exclude>
        </excludes>
      </resource>
    </resources>
    <testResources>
      <testResource>
        <targetPath>string</targetPath>
        <directory>string</directory>
        <includes>
          <include>string</include>
        </includes>
        <excludes>
          <exclude>string</exclude>
        </excludes>
      </testResource>
    </testResources>
    <directory>string</directory>
    <finalName>string</finalName>
    <filters>
      <filter>string</filter>
    </filters>
    <pluginManagement>
      <plugins>
        <plugin>
          <artifactId>maven-compiler-plugin</artifactId>
          <version>${compiler-plugin.version}</version>
          <executions>
            <execution>
              <id>string</id>
              <phase>string</phase>
              <goals>
                <goal>string</goal>
              </goals>
              <inherited>string</inherited>
              <configuration>
                <skip>false</skip>
              </configuration>
            </execution>
          </executions>
          <dependencies>
            <dependency>
              <groupId>string</groupId>
              <artifactId>string</artifactId>
              <version>string</version>
              <type>string</type>
              <classifier>string</classifier>
              <scope>string</scope>
              <systemPath>string</systemPath>
              <exclusions>
                <exclusion>
                  <groupId>string</groupId>
                  <artifactId>string</artifactId>
                </exclusion>
              </exclusions>
              <optional>true</optional>
            </dependency>
          </dependencies>
          <goals>
            <goal>integration-test</goal>
          </goals>
          <inherited>true</inherited>
          <configuration>
            <source>${java.version}</source>
            <target>${java.version}</target>
            <parameters>true</parameters>
          </configuration>
        </plugin>
      </plugins>
    </pluginManagement>
    <plugins>
      <plugin>
        <groupId>org.jacoco</groupId>
        <artifactId>jacoco-maven-plugin</artifactId>
        <version>${jacoco.version}</version>
        <extensions>true</extensions>
        <executions>
          <execution>
            <id>check</id>
            <phase>verify</phase>
            <goals>
              <goal>check</goal>
            </goals>
            <inherited>true</inherited>
            <configuration>
              <dataFile>target/jacoco/allTest.exec</dataFile>
            </configuration>
          </execution>
        </executions>
        <dependencies>
          <dependency>
            <groupId>string</groupId>
            <artifactId>string</artifactId>
            <version>string</version>
            <type>string</type>
            <classifier>string</classifier>
            <scope>string</scope>
            <systemPath>string</systemPath>
            <exclusions>
              <exclusion>
                <groupId>string</groupId>
                <artifactId>string</artifactId>
              </exclusion>
            </exclusions>
          </dependency>
        </dependencies>
        <goals>
          <goal>check</goal>
        </goals>
        <inherited>false</inherited>
        <configuration>
          <rules>
            <rule>
              <element>CLASS</element>
            </rule>
          </rules>
        </configuration>
      </plugin>
    </plugins>
  </build>
  <profiles>
    <profile>
      <id>string</id>
      <activation>
        <activeByDefault>true</activeByDefault>
        <jdk>string</jdk>
        <os>
          <name>string</name>
          <family>string</family>
          <arch>string</arch>
          <version>string</version>
        </os>
        <property>
          <name>string</name>
          <value>string</value>
        </property>
        <file>
          <missing>string</missing>
          <exists>string</exists>
        </file>
      </activation>
      <build>
        <defaultGoal>string</defaultGoal>
        <resources>
          <resource>
            <targetPath>string</targetPath>
            <directory>string</directory>
            <includes>
              <include>string</include>
            </includes>
            <excludes>
              <exclude>string</exclude>
            </excludes>
          </resource>
        </resources>
        <testResources>
          <testResource>
            <targetPath>string</targetPath>
            <filtering>true</filtering>
            <directory>string</directory>
            <includes>
              <include>string</include>
            </includes>
            <excludes>
              <exclude>string</exclude>
            </excludes>
          </testResource>
        </testResources>
        <directory>string</directory>
        <finalName>string</finalName>
        <filters>
          <filter>string</filter>
        </filters>
        <pluginManagement>
          <plugins>
            <plugin>
              <groupId>string</groupId>
              <artifactId>string</artifactId>
              <version>string</version>
              <extensions>true</extensions>
              <executions>
                <execution>
                  <id>string</id>
                  <phase>string</phase>
                  <goals>
                    <goal>string</goal>
                  </goals>
                  <inherited>string</inherited>
                  <configuration>
                    <foo>bar</foo>
                  </configuration>
                </execution>
              </executions>
              <dependencies>
                <dependency>
                  <groupId>string</groupId>
                  <artifactId>string</artifactId>
                  <version>string</version>
                  <type>string</type>
                  <classifier>string</classifier>
                  <scope>string</scope>
                  <systemPath>string</systemPath>
                  <exclusions>
                    <exclusion>
                      <groupId>string</groupId>
                      <artifactId>string</artifactId>
                    </exclusion>
                  </exclusions>
                  <optional>true</optional>
                </dependency>
              </dependencies>
              <goals>
                <foo>bar</foo>
              </goals>
              <inherited>string</inherited>
              <configuration>
                <foo>bar</foo>
              </configuration>
            </plugin>
          </plugins>
        </pluginManagement>
        <plugins>
          <plugin>
            <groupId>string</groupId>
            <artifactId>string</artifactId>
            <version>string</version>
            <executions>
              <execution>
                <id>string</id>
                <phase>string</phase>
                <goals>
                  <goal>string</goal>
                </goals>
                <inherited>string</inherited>
                <configuration>
                  <foo>bar</foo>
                </configuration>
              </execution>
            </executions>
            <dependencies>
              <dependency>
                <groupId>string</groupId>
                <artifactId>string</artifactId>
                <version>string</version>
                <type>string</type>
                <classifier>string</classifier>
                <scope>string</scope>
                <systemPath>string</systemPath>
                <exclusions>
                  <exclusion>
                    <groupId>string</groupId>
                    <artifactId>string</artifactId>
                  </exclusion>
                </exclusions>
              </dependency>
            </dependencies>
            <goals>
              <foo>bar</foo>
            </goals>
            <inherited>string</inherited>
            <configuration>
              <foo>bar</foo>
            </configuration>
          </plugin>
        </plugins>
      </build>
      <modules>
        <module>string</module>
      </modules>
      <repositories>
        <repository>
          <releases>
            <updatePolicy>string</updatePolicy>
            <checksumPolicy>string</checksumPolicy>
          </releases>
          <snapshots>
            <enabled>false</enabled>
            <updatePolicy>string</updatePolicy>
            <checksumPolicy>string</checksumPolicy>
          </snapshots>
          <id>string</id>
          <name>string</name>
          <url>string</url>
          <layout>string</layout>
        </repository>
      </repositories>
      <pluginRepositories>
        <pluginRepository>
          <releases>
            <enabled>false</enabled>
            <updatePolicy>string</updatePolicy>
            <checksumPolicy>string</checksumPolicy>
          </releases>
          <snapshots>
            <updatePolicy>string</updatePolicy>
            <checksumPolicy>string</checksumPolicy>
          </snapshots>
          <id>string</id>
          <name>string</name>
          <url>string</url>
          <layout>string</layout>
        </pluginRepository>
      </pluginRepositories>
      <dependencies>
        <dependency>
          <groupId>string</groupId>
          <artifactId>string</artifactId>
          <version>string</version>
          <type>string</type>
          <classifier>string</classifier>
          <scope>string</scope>
          <systemPath>string</systemPath>
          <exclusions>
            <exclusion>
              <groupId>string</groupId>
              <artifactId>string</artifactId>
            </exclusion>
          </exclusions>
          <optional>true</optional>
        </dependency>
      </dependencies>
      <reports>
        <foo>bar</foo>
      </reports>
      <reporting>
        <outputDirectory>string</outputDirectory>
        <plugins>
          <plugin>
            <groupId>string</groupId>
            <artifactId>string</artifactId>
            <version>string</version>
            <inherited>string</inherited>
            <configuration>
              <foo>bar</foo>
            </configuration>
            <reportSets>
              <reportSet>
                <id>string</id>
                <configuration>
                  <foo>bar</foo>
                </configuration>
                <inherited>string</inherited>
                <reports>
                  <report>string</report>
                </reports>
              </reportSet>
            </reportSets>
          </plugin>
        </plugins>
      </reporting>
      <dependencyManagement>
        <dependencies>
          <dependency>
            <groupId>string</groupId>
            <artifactId>string</artifactId>
            <version>string</version>
            <type>string</type>
            <classifier>string</classifier>
            <scope>string</scope>
            <systemPath>string</systemPath>
            <exclusions>
              <exclusion>
                <groupId>string</groupId>
                <artifactId>string</artifactId>
              </exclusion>
            </exclusions>
            <optional>true</optional>
          </dependency>
        </dependencies>
      </dependencyManagement>
      <distributionManagement>
        <repository>
          <id>string</id>
          <name>string</name>
          <url>string</url>
          <layout>string</layout>
        </repository>
        <snapshotRepository>
          <id>string</id>
          <name>string</name>
          <url>string</url>
          <layout>string</layout>
        </snapshotRepository>
        <site>
          <id>string</id>
          <name>string</name>
          <url>string</url>
        </site>
        <downloadUrl>string</downloadUrl>
        <relocation>
          <groupId>string</groupId>
          <artifactId>string</artifactId>
          <version>string</version>
          <message>string</message>
        </relocation>
        <status>string</status>
      </distributionManagement>
      <properties>
        <prefer.https>true</prefer.https>
      </properties>
    </profile>
  </profiles>
  <modules>
    <module>string</module>
  </modules>
  <repositories>
    <repository>
      <releases>
        <updatePolicy>string</updatePolicy>
        <checksumPolicy>string</checksumPolicy>
      </releases>
      <snapshots>
        <enabled>false</enabled>
        <updatePolicy>string</updatePolicy>
        <checksumPolicy>string</checksumPolicy>
      </snapshots>
      <id>string</id>
      <name>string</name>
      <url>string</url>
      <layout>string</layout>
    </repository>
  </repositories>
  <pluginRepositories>
    <pluginRepository>
      <releases>
        <enabled>false</enabled>
        <updatePolicy>string</updatePolicy>
        <checksumPolicy>string</checksumPolicy>
      </releases>
      <snapshots>
        <updatePolicy>string</updatePolicy>
        <checksumPolicy>string</checksumPolicy>
      </snapshots>
      <id>string</id>
      <name>string</name>
      <url>string</url>
      <layout>string</layout>
    </pluginRepository>
  </pluginRepositories>
  <dependencies>
    <dependency>
      <groupId>string</groupId>
      <artifactId>string</artifactId>
      <version>string</version>
      <type>string</type>
      <classifier>string</classifier>
      <scope>string</scope>
      <systemPath>string</systemPath>
      <exclusions>
        <exclusion>
          <groupId>string</groupId>
          <artifactId>string</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
  </dependencies>
  <reports>
    <report>index</report>
  </reports>
  <reporting>
    <excludeDefaults>true</excludeDefaults>
    <outputDirectory>string</outputDirectory>
    <plugins>
      <plugin>
        <groupId>string</groupId>
        <artifactId>string</artifactId>
        <version>string</version>
        <inherited>string</inherited>
        <configuration>
          <bar>foo</bar>
        </configuration>
        <reportSets>
          <reportSet>
            <id>report1</id>
            <configuration>
              <foo>bar</foo>
            </configuration>
            <inherited>false</inherited>
            <reports>
              <report>foo</report>
            </reports>
          </reportSet>
        </reportSets>
      </plugin>
    </plugins>
  </reporting>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>string</groupId>
        <artifactId>string</artifactId>
        <version>string</version>
        <type>string</type>
        <classifier>string</classifier>
        <scope>string</scope>
        <systemPath>string</systemPath>
        <exclusions>
          <exclusion>
            <groupId>string</groupId>
            <artifactId>string</artifactId>
          </exclusion>
        </exclusions>
        <optional>true</optional>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <distributionManagement>
    <repository>
      <id>string</id>
      <name>string</name>
      <url>string</url>
      <layout>string</layout>
    </repository>
    <snapshotRepository>
      <id>string</id>
      <name>string</name>
      <url>string</url>
      <layout>string</layout>
    </snapshotRepository>
    <site>
      <id>string</id>
      <name>string</name>
      <url>string</url>
    </site>
    <downloadUrl>string</downloadUrl>
    <relocation>
      <groupId>string</groupId>
      <artifactId>string</artifactId>
      <version>string</version>
      <message>string</message>
    </relocation>
    <status>string</status>
  </distributionManagement>
  <properties>
    <maven.compiler.parameters>true</maven.compiler.parameters>
    <maven.javadoc.failOnError>false</maven.javadoc.failOnError>
  </properties>
</project>
